core/server/data/migrations/versions/5.0/2022-03-28-15-25-backfill-mrr-adjustments-for-offers.js

```diff
@@ -0,0 +1,103 @@
+const logging = require('@tryghost/logging');
+const {uniq} = require('lodash');
+
+const {createTransactionalMigration} = require('../../utils');
+
+module.exports = createTransactionalMigration(
+    async function up(knex) {
+        logging.info('Adjusting MRR based on Offer Redemptions');
+        const offerRedemptions = await knex
+            .select('or.*', 'o.discount_type', 'o.discount_amount', 'o.interval AS discount_interval', 's.mrr AS mrr', 'p.amount AS amount', 'p.interval AS interval')
+            .from('offer_redemptions AS or')
+            .join('offers AS o', 'or.offer_id', '=', 'o.id')
+            .join('members_stripe_customers_subscriptions AS s', 'or.subscription_id', '=', 's.id')
+            .join('stripe_prices AS p', 'p.stripe_price_id', '=', 's.stripe_price_id')
+            .where('o.duration', '=', 'forever')
+            .whereNotNull('s.offer_id')
+            .where('s.mrr', '!=', 0)
+            .orderBy('or.created_at', 'asc');
+
+        if (offerRedemptions.length === 0) {
+            logging.info('No Offers redeemed, skipping migration');
+            return;
+        } else {
+            logging.info(`Adjusting MRR for ${offerRedemptions.length} Offer Redemptions`);
+        }
+
+        const memberIds = uniq(offerRedemptions.map(redemption => redemption.member_id));
+
+        const mrrEvents = await knex
+            .select('*')
+            .from('members_paid_subscription_events')
+            .whereIn('member_id', memberIds);
+
+        function storeEventOnMemberId(storage, event) {
+            return {
+                ...storage,
+                [event.member_id]: storage[event.member_id] ? storage[event.member_id].concat(event) : [event]
+            };
+        }
+
+        const mrrEventsByMemberId = mrrEvents.reduce(storeEventOnMemberId, {});
+
+        const updatedEvents = [];
+
+        function calculateMRR(event, redemption) {
+            if (redemption && event.interval !== redemption.discount_interval) {
+                logging.error('Found invalid price & redemption pair');
+                return calculateMRR(event);
+            }
+
+            if (!redemption) {
+                return event.interval === 'year' ? event.amount / 12 : event.amount;
+            }
+
+            if (redemption.discount_type === 'percent') {
+                return calculateMRR({
+                    interval: event.interval,
+                    amount: event.amount * (100 - redemption.discount_amount) / 100
+                });
+            }
+
+            return calculateMRR({
+                interval: event.interval,
+                amount: event.amount - redemption.discount_amount
+            });
+        }
+
+        offerRedemptions.forEach((redemption) => {
```

Comment:
I thought this would need an eslint-ignore line 🤔

Reply:
Doesn't seem to! maybe `forEach` isn't in the linting